### ISSUE TITLE
Vectors serialization error

### ISSUE THREAD

**Issue by sanyabt** [clulab/eidos#884]
Hi

I am trying to use the Eidos reader (as a webapp) with grounding enabled and using the MESH ontology. After making the changes in eidos.conf (as mentioned [here](https://github.com/clulab/eidos#configuring)) i.e. specifying useGrounding=True and adding "mesh" to the list of ontologies, I followed the instructions to cache the ontologies (as [here](https://github.com/clulab/eidos#optimizing)). However, I run into errors every time. I've attached part of the terminal output below:

`[info] Packaging /Users/Seema/eidos/target/scala-2.12/eidos_2.12-1.1.0-SNAPSHOT.jar ...
[info] Done packaging.
[info] Running (fork) org.clulab.wm.eidos.apps.CacheOntologies 
[info] Saving vectors to ./cache/english/glove.840B.300d.txt.serialized...
[error] 23:45:38.341 [main] INFO  o.c.wm.eidos.EidosComponentsBuilder$ - Loading config...
[error] 23:45:38.469 [scala-execution-context-global-9] INFO  o.c.wm.eidos.EidosComponentsBuilder$ - Loading processor...
[error] 23:45:38.508 [scala-execution-context-global-9] INFO  org.clulab.wm.eidos.utils.Sourcer$ - Sourcing resource file:/private/var/folders/73/1n5gwsnx5dl08jq_9mf6cvnc0000gn/T/sbt_de2454aa/job-1/target/c242e1c3/eidos_2.12-1.1.0-SNAPSHOT.jar!/org/clulab/wm/eidos/english/filtering/stops.txt
[error] 23:45:38.536 [scala-execution-context-global-9] INFO  org.clulab.wm.eidos.utils.Sourcer$ - Sourcing resource file:/private/var/folders/73/1n5gwsnx5dl08jq_9mf6cvnc0000gn/T/sbt_de2454aa/job-1/target/c242e1c3/eidos_2.12-1.1.0-SNAPSHOT.jar!/org/clulab/wm/eidos/english/filtering/transparent.txt
[error] 23:45:38.565 [scala-execution-context-global-12] INFO  org.clulab.wm.eidos.utils.Sourcer$ - Sourcing resource file:/private/var/folders/73/1n5gwsnx5dl08jq_9mf6cvnc0000gn/T/sbt_de2454aa/job-1/target/c242e1c3/eidos_2.12-1.1.0-SNAPSHOT.jar!/org/clulab/wm/eidos/english/quantifierKB/gradable_adj_fullmodel.kb
[error] 23:45:38.565 [scala-execution-context-global-9] INFO  o.c.w.e.groundings.OntologyHandler$ - Loading W2V...
[error] 23:45:38.569 [scala-execution-context-global-9] INFO  o.c.w.e.groundings.EidosWordToVec$ - Loading w2v from /org/clulab/glove/glove.840B.300d.txt...
[error] 23:45:38.579 [scala-execution-context-global-9] TRACE o.c.e.word2vec.CompactWord2Vec - Started to load word2vec matrix from file /org/clulab/glove/glove.840B.300d.txt...
[error] 23:45:38.584 [scala-execution-context-global-9] INFO  org.clulab.utils.Sourcer$ - Sourcing resource file:/private/var/folders/73/1n5gwsnx5dl08jq_9mf6cvnc0000gn/T/sbt_de2454aa/target/0d9e1a32/glove-840b-300d-0.1.0.jar!/org/clulab/glove/glove.840B.300d.txt
[error] 23:45:38.700 [scala-execution-context-global-10] INFO  org.clulab.sequences.LexiconNER - Beginning to load the KBs for the rule-based bio NER...
[error] 23:45:38.700 [scala-execution-context-global-12] INFO  org.clulab.wm.eidos.utils.Sourcer$ - Sourcing resource file:/private/var/folders/73/1n5gwsnx5dl08jq_9mf6cvnc0000gn/T/sbt_de2454aa/job-1/target/c242e1c3/eidos_2.12-1.1.0-SNAPSHOT.jar!/org/clulab/wm/eidos/english/grammars/taxonomy.yml
[error] 23:45:38.709 [scala-execution-context-global-11] INFO  org.clulab.wm.eidos.utils.Sourcer$ - Sourcing resource file:/private/var/folders/73/1n5gwsnx5dl08jq_9mf6cvnc0000gn/T/sbt_de2454aa/job-1/target/c242e1c3/eidos_2.12-1.1.0-SNAPSHOT.jar!/org/clulab/wm/eidos/english/confidence/hedging.txt
[error] 23:45:38.822 [scala-execution-context-global-10] INFO  org.clulab.sequences.LexiconNER - Loaded matcher for label Quantifier. The number of entries added to the first layer was 221.
[error] 23:45:38.882 [scala-execution-context-global-10] INFO  org.clulab.sequences.LexiconNER - Loaded matcher for label Property. The number of entries added to the first layer was 159.
[error] 23:45:38.901 [scala-execution-context-global-10] INFO  org.clulab.sequences.LexiconNER - Loaded matcher for label Location. The number of entries added to the first layer was 218.
[error] 23:45:38.902 [scala-execution-context-global-10] INFO  org.clulab.sequences.LexiconNER - KB loading completed.
[error] 23:45:38.975 [scala-execution-context-global-10] INFO  org.clulab.wm.eidos.utils.Sourcer$ - Sourcing resource file:/private/var/folders/73/1n5gwsnx5dl08jq_9mf6cvnc0000gn/T/sbt_de2454aa/job-1/target/c242e1c3/eidos_2.12-1.1.0-SNAPSHOT.jar!/org/clulab/wm/eidos/english/grammars/entities/grammar/entities.yml
[error] 23:45:42.221 [scala-execution-context-global-10] INFO  org.clulab.wm.eidos.utils.Sourcer$ - Sourcing resource file:/private/var/folders/73/1n5gwsnx5dl08jq_9mf6cvnc0000gn/T/sbt_de2454aa/job-1/target/c242e1c3/eidos_2.12-1.1.0-SNAPSHOT.jar!/org/clulab/wm/eidos/english/grammars/avoidLocal.yml
[error] 2020-06-29 23:45:51.973732: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.2 AVX AVX2 FMA
[error] 23:45:52.812 [scala-execution-context-global-10] INFO  org.clulab.wm.eidos.utils.Sourcer$ - Sourcing resource file:/private/var/folders/73/1n5gwsnx5dl08jq_9mf6cvnc0000gn/T/sbt_de2454aa/job-1/target/c242e1c3/eidos_2.12-1.1.0-SNAPSHOT.jar!/org/clulab/wm/eidos/english/grammars/context.yml
[error] 23:45:52.988 [scala-execution-context-global-10] INFO  org.clulab.wm.eidos.utils.Sourcer$ - Sourcing resource file:/private/var/folders/73/1n5gwsnx5dl08jq_9mf6cvnc0000gn/T/sbt_de2454aa/job-1/target/c242e1c3/eidos_2.12-1.1.0-SNAPSHOT.jar!/org/clulab/wm/eidos/english/grammars/master.yml
...`

Running on macOS Mojave 10.14.5 with latest version of Eidos repository. I've been able to make the webapp run with a previous Eidos version but can't move ahead with this. Any help would be appreciated. Thanks!

**Comment by bgyori** [comment]
Hi @sanyabt, while I can't help with this specific technical issue, I have some - hopefully useful - thoughts about this. The MeSH ontology was added to Eidos as an experiment to see if the embeddings-based grounding model (which builds on the names and examples or descriptions associated with ontology entries to find matches) could work with a biomedical ontology. We found this approach to be noisy but perhaps further improvements could be made to make it work well. Since I know you're also using INDRA, I wanted to point you to an alternative, which seems to fare better in terms of grounding accuracy. This approach takes ungrounded Eidos output (i.e., no need to configure MeSH here) and uses Gilda (https://github.com/indralab/gilda) via INDRA to ground normalize entity texts picked up as arguments of causal relations by Eidos to biomedical ontologies including MeSH. See these two API calls in INDRA: https://indra.readthedocs.io/en/latest/modules/sources/eidos/index.html#indra.sources.eidos.api.process_json_bio, https://indra.readthedocs.io/en/latest/modules/sources/eidos/index.html#indra.sources.eidos.api.process_text_bio, in case you're interested in this alternative workflow, I'm happy to follow up.

**Comment by sanyabt** [comment]
Hi @bgyori, thank you for the suggestion! That's actually pretty interesting and yes, we've been working with INDRA (and REACH) for biomedical literature machine reading. REACH is extracting the mechanistic statements pretty well but we were hoping to leverage EIDOS for the general purpose statements (such as in observational studies) - would it then be possible to ground the entities through INDRA's bio_ontology with the API calls you mentioned or with https://indra.readthedocs.io/en/latest/modules/sources/eidos/index.html#indra.sources.eidos.api.reground_texts? So the interest would be more in recognizing the biomedical entities through EIDOS (in the open domain statements) rather than inclusion of additional Activation/Inhibition statements. 

Additionally, are there any results I could look at for the grounding accuracy of various readers (and ontologies) for various texts? Would be very helpful in determining the best processes to follow in our pipeline. (I admit I'm still going through your paper on INDRA's implementation and use cases). Thank you again for all your help and prompt replies, I really appreciate it. INDRA is such a powerful framework and my coworker and I have been enjoying exploring all the modules.

(Let me know if I should close this issue since we diverged from the technical discussion quite a bit)

**Comment by kwalcock** [comment]
I'm not sure why we are logging to [error].  Perhaps it is to distinguish our messages from what sbt otherwise outputs.  The ones you pasted do not yet indicate an error.  They are all INFO and TRACE statements.  Do you have more?  I do get out of memory errors and am checking that out.

```
[info] Packaging /Users/Seema/eidos/target/scala-2.12/eidos_2.12-1.1.0-SNAPSHOT.jar ... 
[info] Done packaging. 
[info] Running (fork) org.clulab.wm.eidos.apps.CacheOntologies 
[info] Saving vectors to ./cache/english/glove.840B.300d.txt.serialized... 
[error] 23:45:38.341 [main] INFO o.c.wm.eidos.EidosComponentsBuilder$ - Loading config... 
[error] 23:45:38.469 [scala-execution-context-global-9] INFO o.c.wm.eidos.EidosComponentsBuilder$ - Loading processor... 
[error] 23:45:38.508 [scala-execution-context-global-9] INFO org.clulab.wm.eidos.utils.Sourcer$ - Sourcing resource file:/private/var/folders/73/1n5gwsnx5dl08jq_9mf6cvnc0000gn/T/sbt_de2454aa/job-1/target/c242e1c3/eidos_2.12-1.1.0-SNAPSHOT.jar!/org/clulab/wm/eidos/english/filtering/stops.txt 
[error] 23:45:38.536 [scala-execution-context-global-9] INFO org.clulab.wm.eidos.utils.Sourcer$ - Sourcing resource file:/private/var/folders/73/1n5gwsnx5dl08jq_9mf6cvnc0000gn/T/sbt_de2454aa/job-1/target/c242e1c3/eidos_2.12-1.1.0-SNAPSHOT.jar!/org/clulab/wm/eidos/english/filtering/transparent.txt 
[error] 23:45:38.565 [scala-execution-context-global-12] INFO org.clulab.wm.eidos.utils.Sourcer$ - Sourcing resource file:/private/var/folders/73/1n5gwsnx5dl08jq_9mf6cvnc0000gn/T/sbt_de2454aa/job-1/target/c242e1c3/eidos_2.12-1.1.0-SNAPSHOT.jar!/org/clulab/wm/eidos/english/quantifierKB/gradable_adj_fullmodel.kb 
[error] 23:45:38.565 [scala-execution-context-global-9] INFO o.c.w.e.groundings.OntologyHandler$ - Loading W2V... 
[error] 23:45:38.569 [scala-execution-context-global-9] INFO o.c.w.e.groundings.EidosWordToVec$ - Loading w2v from /org/clulab/glove/glove.840B.300d.txt... 
[error] 23:45:38.579 [scala-execution-context-global-9] TRACE o.c.e.word2vec.CompactWord2Vec - Started to load word2vec matrix from file /org/clulab/glove/glove.840B.300d.txt... 
[error] 23:45:38.584 [scala-execution-context-global-9] INFO org.clulab.utils.Sourcer$ - Sourcing resource file:/private/var/folders/73/1n5gwsnx5dl08jq_9mf6cvnc0000gn/T/sbt_de2454aa/target/0d9e1a32/glove-840b-300d-0.1.0.jar!/org/clulab/glove/glove.840B.300d.txt 
[error] 23:45:38.700 [scala-execution-context-global-10] INFO org.clulab.sequences.LexiconNER - Beginning to load the KBs for the rule-based bio NER... 
[error] 23:45:38.700 [scala-execution-context-global-12] INFO org.clulab.wm.eidos.utils.Sourcer$ - Sourcing resource file:/private/var/folders/73/1n5gwsnx5dl08jq_9mf6cvnc0000gn/T/sbt_de2454aa/job-1/target/c242e1c3/eidos_2.12-1.1.0-SNAPSHOT.jar!/org/clulab/wm/eidos/english/grammars/taxonomy.yml 
[error] 23:45:38.709 [scala-execution-context-global-11] INFO org.clulab.wm.eidos.utils.Sourcer$ - Sourcing resource file:/private/var/folders/73/1n5gwsnx5dl08jq_9mf6cvnc0000gn/T/sbt_de2454aa/job-1/target/c242e1c3/eidos_2.12-1.1.0-SNAPSHOT.jar!/org/clulab/wm/eidos/english/confidence/hedging.txt 
[error] 23:45:38.822 [scala-execution-context-global-10] INFO org.clulab.sequences.LexiconNER - Loaded matcher for label Quantifier. The number of entries added to the first layer was 221. 
[error] 23:45:38.882 [scala-execution-context-global-10] INFO org.clulab.sequences.LexiconNER - Loaded matcher for label Property. The number of entries added to the first layer was 159. 
[error] 23:45:38.901 [scala-execution-context-global-10] INFO org.clulab.sequences.LexiconNER - Loaded matcher for label Location. The number of entries added to the first layer was 218. 
[error] 23:45:38.902 [scala-execution-context-global-10] INFO org.clulab.sequences.LexiconNER - KB loading completed. 
[error] 23:45:38.975 [scala-execution-context-global-10] INFO org.clulab.wm.eidos.utils.Sourcer$ - Sourcing resource file:/private/var/folders/73/1n5gwsnx5dl08jq_9mf6cvnc0000gn/T/sbt_de2454aa/job-1/target/c242e1c3/eidos_2.12-1.1.0-SNAPSHOT.jar!/org/clulab/wm/eidos/english/grammars/entities/grammar/entities.yml 
[error] 23:45:42.221 [scala-execution-context-global-10] INFO org.clulab.wm.eidos.utils.Sourcer$ - Sourcing resource file:/private/var/folders/73/1n5gwsnx5dl08jq_9mf6cvnc0000gn/T/sbt_de2454aa/job-1/target/c242e1c3/eidos_2.12-1.1.0-SNAPSHOT.jar!/org/clulab/wm/eidos/english/grammars/avoidLocal.yml 
[error] 2020-06-29 23:45:51.973732: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.2 AVX AVX2 FMA 
[error] 23:45:52.812 [scala-execution-context-global-10] INFO org.clulab.wm.eidos.utils.Sourcer$ - Sourcing resource file:/private/var/folders/73/1n5gwsnx5dl08jq_9mf6cvnc0000gn/T/sbt_de2454aa/job-1/target/c242e1c3/eidos_2.12-1.1.0-SNAPSHOT.jar!/org/clulab/wm/eidos/english/grammars/context.yml 
[error] 23:45:52.988 [scala-execution-context-global-10] INFO org.clulab.wm.eidos.utils.Sourcer$ - Sourcing resource file:/private/var/folders/73/1n5gwsnx5dl08jq_9mf6cvnc0000gn/T/sbt_de2454aa/job-1/target/c242e1c3/eidos_2.12-1.1.0-SNAPSHOT.jar!/org/clulab/wm/eidos/english/grammars/master.yml ...
```

**Comment by kwalcock** [comment]
The memory problem was exacerbated by me using an experimental branch and it's working better in master.  MeSH will require lots of memory.  I gave it _JAVA_OPTIONS: -Xmx24g, but it's using about 14GB right now (perhaps unnecessarily).  I also commented out all lines at the bottom of CacheOntologies.scala except `safeCacheOntologies()` so that it wouldn't try to do all of the memory intensive operations in one run.  `replaceGeoNorms()` and `cacheWord2Vec()` can be done in separate passes.  You might want to change the setting for `includeParents` in eidos.conf to `false` for MeSH.  That's something that was added later that doesn't especially apply to MeSH.  These are all exceptions to the normal procedure because this ontology is so large.
